Rhys/main.cpp

```diff
@@ -0,0 +1,195 @@
+    if (!file) {
+        cout << "Failed to open user file." << endl;
+        return;
+    }
+    while (getline(file, entryLine)) {
+        stringstream ss(entryLine);
+        User user;
+        string friends, friendName;
+        getline(ss, user.username, ';');
+        getline(ss, user.displayName, ';');
+        getline(ss, user.state, ';');
+        getline(ss, friends);
+        friends = friends.substr(1, friends.length() - 2);
+
+        stringstream fs(friends);
+        while (getline(fs, friendName, ',')) {
+            if (!friendName.empty()) {
+                user.friends.push_back(friendName);
+            }
+        }
+        users.push_back(user);
+    }
+    dataLoaded = true;  // Set data loaded flag to true after successful load
+}
+
+void loadPostData(string filename){
+    ifstream file(filename);
+    string entryLine;
+    if (!file) {
+        cout << "Failed to open post file." << endl;
+        return;
+    }
+    while(getline(file, entryLine)){
+        stringstream ss(entryLine);
+        Post post;
+        getline(ss, post.postId, ';');
+        getline(ss, post.userId, ';');
+        getline(ss, post.visibility);
+        posts.push_back(post);
+    }
+}
+
+string checkVisibility(string postId, string username){
+    if (!dataLoaded) {
+        return "Please load data first";
+    }
+    for(const Post& post : posts){
+        if(post.postId == postId){
+            if(post.visibility == "public"){
+                return "Access Permitted";
+            } else {
+                // Check if the username is in the friends list of the post's user
+                auto it = find_if(users.begin(), users.end(), [&post](const User& u) {
+                    return u.username == post.userId;
+                });
+                if (it != users.end() && find(it->friends.begin(), it->friends.end(), username) != it->friends.end()) {
+                    return "Access Permitted";
+                }
+            }
+        }
+    }
+    return "Access Denied";
+}
+
+vector<string> retrievePosts(string username) {
+    if (!dataLoaded) {
+        return {"Please load data first"};
+    }
+    vector<string> visiblePosts;
+    for (const Post& post : posts) {
+        if (post.userId != username || 
+            (post.visibility == "friend" && find_if(users.begin(), users.end(), [&post, &username](const User& u) {
+                return u.username == post.userId && find(u.friends.begin(), u.friends.end(), username) != u.friends.end();
+            }) != users.end())) {
+            visiblePosts.push_back(post.postId);
+        }
+    }
+    return visiblePosts;
+}
+
+vector<string> searchUsersByLocation(string state){
+    if (!dataLoaded) {
+        return {"Please load data first"};
+    }
+    vector<string> matchedUsers;
+    for(const User& user : users){
+        if(user.state == state){
+            matchedUsers.push_back(user.displayName);
+        }
+    }
+    return matchedUsers;
+}
+
+int main(){

```

Comment:
*Main*
+Looks very pretty because you created methods for everything. 
+Very easy to read because its nice and simple.